### ISSUE TITLE
Enum API

### ISSUE THREAD

**Issue by Hosch250** [dotnet/corefx#38822]
Now that we have the `where T : enum` flag, would it be possible to implement a generic version of the `Enum` API?

```cs
public string GetName<T>(T value) where T : enum
    => GetName(typeof(T), value);

public string[] GetNames<T>() where T : enum
    => GetNames(typeof(T));

public T[] GetValues<T>() where T : enum
    => GetValues(typeof(T));

public T Parse<T>(T value) where T : enum
    => Parse(typeof(T), value);

public object ToObject<T>(T value) where T : enum
    => ToObject(typeof(T), value);
```

Note that `TryParse` already uses generics, so it would make sense to update the rest of the API. Also, `GetValues` would benefit from being a typed array instead of the old non-generic `Array`.

**Comment by joperezr** [comment]
@Hosch250 Are you interested in submitting a formal proposal? The link provided above by @RussKeldorph should have all the info you need.

**Comment by Hosch250** [comment]
@joperezr Sure. I'll take a look right now.

**Comment by Hosch250** [comment]
The `System.Enum` class implements an API that would have readability, type safety, and performance enhanced by implementing the methods with generics. Now that .NET Core supports the `Enum` filter on generic type members it should be easy to implement.

# Rational and Usage
Currently, to get the name of an enum value, you must do:

```cs
MyEnum value;
Enum.GetName(typeof(MyEnum), value);
```

Or to get all values of an enum type, you must do:

```cs
var values = Enum.GetValues(typeof(MyEnum)).Cast<MyEnum>();
```

With generics, we could make these calls more clear and concise as:

```cs
MyEnum value;
Enum.GetName<MyEnum>(value);

var values = Enum.GetValues<MyEnum>();
```

# Proposed API
I propose we add the following members to `System.Enum`:

```cs
public static string GetName<T>(T value) where T : enum;
public static IEnumerable<string> GetNames<T>() where T : enum;
public static IEnumerable<T> GetValues<T>() where T : enum;
public static T Parse<T>(T value) where T : enum;
public static object ToObject<T>(T value) where T : enum;
```

# Open Questions

Should the `IEnumerable` members be `IList` or `Array<T>` instead?

**Comment by Hosch250** [comment]
@joperezr Does this work? Should I move it to the top of the description?

**Comment by mburbea** [comment]
just a note 
`var values = Enum.GetValues(typeof(MyEnum)).Cast<MyEnum>();` is not needed.
The result of `Enum.GetValues(typeof(MyEnum))` is a `MyEnum[]`, you just need to do this::
`var values = (MyEnum[])Enum.GetValues(typeof(MyEnum))`

You can also cast it to it's underlying type leading to this strange situation:
```cs
var array =(int[]) Enum.GetValues(typeof(DayOfWeek);
Console.Write(array[1].GetType() != array.GetValue(1).GetType()); // prints true
```

**Comment by joperezr** [comment]
> Does this work? Should I move it to the top of the description?

Yeah that looks good, and yes please move it to the top of the description.

**Comment by Hosch250** [comment]
@mburbea Good points. But it's not easy to see that with intellisense--it just says the type is `Array`.
